### ISSUE TITLE
Add feture to check the status of NASA APIs

### ISSUE THREAD

**Issue by betolink** [nsidc/earthdata#161]
Currently there is no way of checking the status of the APIs used by the library, it would be useful having a `status()` function (maybe in the module __init__) to verify that CMR and EDL are working and available.

**Comment by asteiker** [comment]
This [zulip message](https://earthaccess.zulipchat.com/#narrow/channel/480805-help/topic/Ask.20anything.20HERE!/near/517934383) is related to this feature. @mfisher87 notes that there is a public status.earthdata.nasa.gov endpoint, though I don't know if/how this could be inspected programmatically to determine if a CMR outage is impacting an earthaccess user experience. If that's possible, it would be a great improvement!

**Comment by Sherwin-14** [comment]
@mfisher87 @chuckwondo  One possible solution(although this is a naive one I feel) would be to scrape the page Amy linked above, and pick out details that are of our interest. Although, if the page structure changes, then the function will be rendered useless and that is what I am worried about :(,  What are your opinions on this?

**Comment by chuckwondo** [comment]
We should absolutely _not_ be scraping the status page. In my opinion, we shouldn't be doing any sort of status check at any point in our code. Instead, what we might consider is simply referring the user to go to <https://status.earthdata.nasa.gov/> to check the status of services, if we detect that we've received any 5xx response.

**Comment by Sherwin-14** [comment]
>Instead, what we might consider is simply referring the user to go to https://status.earthdata.nasa.gov/ to check the status of services, if we detect that we've received any 5xx response.

So, if a download fails then we can just return this link to the user so that they can inspect whether the service is down?

**Comment by chuckwondo** [comment]
> > Instead, what we might consider is simply referring the user to go to https://status.earthdata.nasa.gov/ to check the status of services, if we detect that we've received any 5xx response.
> 
> So, if a download fails then we can just return this link to the user so that they can inspect whether the service is down?

Not quite. If any requests that hit the CMR or Earthdata Login fail with a 5xx response, then we raise with an error message that instructs the user to check the status of NASA Earthdata services at that URL.

Opening and downloading are not directly related to NASA Earthdata services. Once a CMR search is done, and proper credentials are obtained, opening and downloading deal either with AWS (for cloud-hosted files) or a DAAC on-prem server, not with NASA Earthdata.

Therefore, only CMR queries and logging in have anything to do with NASA Earthdata, so only when we get a 5xx response from either of these operations do we want to instruct the user to check the services statuses at the URL above.

**Comment by Sherwin-14** [comment]
Thanks for clarifying this! So we need to focus on CMR queries and login functionality and if they return 5xx response then we can return the link to the user. I am planning to work on this, do you have any specific opinions about this?

**Comment by mfisher87** [comment]
There is actually a JSON status API:


```bash
$ curl -s --request GET --url https://status.earthdata.nasa.gov/api/v1/statuses | jq .
{
  "success": true,
  "statuses": [
    {
      "name": "MMT (SIT)",
      "status": "OK"
    },
    {
      "name": "Status App Test Application",
      "status": "OK"
    },
    {
      "name": "NSIDC DAAC OpenAltimetry (UAT)",
      "status": "OUTAGE"
    },
    {
      "name": "ED Pub SIT",
      "status": "OUTAGE"
    },
    {
      "name": "Earthdata Login",
      "status": "OK"
    },
    {
      "name": "Common Metadata Repository (PROD)",
      "status": "OK"
    }
  ]
}

```

I think it'd be pretty cool to let people see nicely-formatted statuses for these services from Python (and in the future, CLI)! 

>In my opinion, we shouldn't be doing any sort of status check at any point in our code

I don't think we should be doing this _implicitly_ (e.g. checking status before doing a search), but if the user wants to write their program to be status-aware, I think it would be valuable to offer them a method they can explicitly call to use status information in their program.

I also like your idea about prompting the user to check the status GUI (e.g. in a error log message) if one of these services 500s. I think these could/should be implemented independently.

**Comment by chuckwondo** [comment]
Yes! I like this. I figured there would be an API behind the status page, but didn't realize that it's publicly available. Nice find!

Correct, no _implicit_ checks.

I like the idea of adding a function to make the call you show above. We might want to return only the statuses of "Earthdata Login" and "Common Metadata Repository (PROD)". Perhaps calling it `earthaccess.status`, as @betolink originally suggested, where it might return a dict like so:

```json
{"Earthdata Login": "OK", "Common Metadata Repository (PROD)": "OK"}
```

**Comment by Sherwin-14** [comment]
So we just need to output a dictionary(similar to what Chuck showed above) when a user calls `earthaccess.status()` function. Should we include this function in the `api.py` or should we add this in `__init__.py` as Luis suggested intially?

**Comment by mfisher87** [comment]
It would be a top-level API function, so I think defining it in `api.py` and importing it in `__init__.py` makes sense.

**Comment by chuckwondo** [comment]
For completeness, which endpoint we query should depend upon which "system" the user is using:

- `earthaccess.PROD` (default): https://status.earthdata.nasa.gov/api/v1/statuses
- `earthaccess.UAT`: https://status.uat.earthdata.nasa.gov/api/v1/statuses

Note that the UAT endpoint produces JSON similar to the following:

```json
{
  "success": true,
    "statuses": [
      {
        "name": "Earthdata Login (UAT)",
        "status": "OK"
      },
      {
        "name": "Test App",
        "status": "OK"
      },
      {
        "name": "Common Metadata Repository",
        "status": "OK"
      }
    ]
}
```

Note the slight differences in the 2 keys we're interested in:

- PROD:
   - "Earthdata Login"
   - "Common Metadata Repository (PROD)"
- UAT
  - "Earthdata Login (UAT)"
  - "Common Metadata Repository"

The differences seem a bit oddly inconsistent, so I propose that in either case, we look for keys that _start with_ "Earthdata Login" and "Common Metadata Repository" and that we return an object of the following form, regardless of system, for consistency (i.e., no `" (PROD)"` nor `" (UAT)"` suffix in the keys):

```json
{"Earthdata Login": "OK", "Common Metadata Repository": "OK"}
```

**Comment by mfisher87** [comment]
>The differences seem a bit oddly inconsistent

I noticed that too, it makes me worry that these strings are likely to change! Hopefully they will version up the API if/when they do that.
